go/database/mpt/verification_test.go

```diff
@@ -668,20 +672,77 @@ func TestVerification_VerificationObserverReportsError(t *testing.T) {
 			node.info.Balance = amount.Add(node.info.Balance, amount.New(1))
 		})
 
-		if err := VerifyMptState(dir, config, roots, observer); err == nil {
+		if err := VerifyMptState(context.Background(), dir, config, roots, observer); err == nil {
 			t.Errorf("Modified node should have been detected")
 		}
 	})
 }
 
 func TestVerification_VerifyValidMptState(t *testing.T) {
 	runVerificationTest(t, func(t *testing.T, dir string, config MptConfig, roots []Root) {
-		if err := VerifyMptState(dir, config, roots, NilVerificationObserver{}); err != nil {
+		if err := VerifyMptState(context.Background(), dir, config, roots, NilVerificationObserver{}); err != nil {
 			t.Errorf("found unexpected error in fresh mpt: %v", err)
 		}
 	})
 }
 
+func TestVerification_CanInterrupt(t *testing.T) {
+	type verifyFunc func(context.Context, string, MptConfig, []Root, VerificationObserver) error
+	tests := map[string]verifyFunc{"VerifyMptState": VerifyMptState, "VerifyFileForest": verifyFileForest}
+
+	for name, verify := range tests {
+		runVerificationTest(t, func(t *testing.T, dir string, config MptConfig, roots []Root) {
+			// First count the occurrence
+			counter := &interruptObserver{t: t}
+			if err := verifyFileForest(context.Background(), dir, config, roots, counter); err != nil {
+				t.Errorf("%v: found unexpected error in fresh forest: %v", name, err)
+			}
+			maxCount := counter.count
+			ctx := interrupt.CancelOnInterrupt(context.Background())
+			// Then interrupt and check it happened
+			interrupter := &interruptObserver{t: t, signalInterrupt: true}
+			if got, want := verify(ctx, dir, config, roots, interrupter), interrupt.ErrCanceled; !errors.Is(got, want) {
+				t.Errorf("%v: unexpected error: got: %v, want: %v", name, got, want)
+			}
+			if interrupter.count >= maxCount {
+				t.Fatalf("%v: verification was not interrupted", name)
+			}
+
+			// Make sure data is not corrupted
+			ns, err := openVerificationNodeSource(nil, dir, config)
+			if err != nil {
+				t.Fatalf("%v: failed open node source: %v", name, err)
+			}
+
+			err = ns.Close()
+			if err != nil {
+				t.Fatalf("%v: failed to close node source: %v", name, err)
+			}
+		})
+	}
+}
+
+type interruptObserver struct {
+	t               *testing.T
+	signalInterrupt bool
+	count           int
+}
+
+func (c *interruptObserver) StartVerification() {
```

Comment:
I do not very like how the interrupt is triggered here. First I think that syscall is asynchronous, and thus this test may be fragile. Second and most importantly, interrupt is triggered based on the number of execution of `StartVerification`. It is not obvious what is the relation between the number of execution of this method and the number of call to the `IsCancelled` method. 

I think much better is to really track the number of execution to `IsCancelled` first, and then simulate a cancelled context in next loops. For idea how to do it, please have a look here:

https://github.com/Fantom-foundation/Carmen/pull/947/files#diff-8bb57076b2dc35d3c0b28ecd9488c8c002f0408e571713e8462c7b7e8a6d06aeR640

Reply:
somehow the link does not lead to the line number I wanted to link. Please search for this line

`type countingWhenDoneContext struct {`